File: abxbus-go/tests/eventbus_serialization_test.go
package abxbus_test

import (
	"bytes"
	"context"
	"encoding/json"
	"testing"

	abxbus "github.com/ArchiveBox/abxbus/abxbus-go/v2"
)

func assertJSONKeyBefore(t *testing.T, data []byte, firstKey string, secondKey string) {
	t.Helper()
	firstNeedle := append(append([]byte{'"'}, []byte(firstKey)...), []byte{'"', ':'}...)
	secondNeedle := append(append([]byte{'"'}, []byte(secondKey)...), []byte{'"', ':'}...)
	firstIndex := bytes.Index(data, firstNeedle)
	secondIndex := bytes.Index(data, secondNeedle)
	if firstIndex < 0 || secondIndex < 0 {
		t.Fatalf("expected JSON keys %q and %q in payload: %s", firstKey, secondKey, string(data))
	}
	if firstIndex > secondIndex {
		t.Fatalf("expected JSON key %q before %q in payload: %s", firstKey, secondKey, string(data))
	}
}

func TestEventBusSerializationRoundtripPreservesConfigHandlersHistory(t *testing.T) {
	maxHistory := 5
	eventTimeout := 2.5
	eventSlowTimeout := 0.75
	handlerSlowTimeout := 0.33
	bus := abxbus.NewEventBus("SerBus", &abxbus.EventBusOptions{
		ID:                      "serbus-1234",
		MaxHistorySize:          &maxHistory,
		MaxHistoryDrop:          true,
		EventConcurrency:        abxbus.EventConcurrencyParallel,
		EventTimeout:            &eventTimeout,
		EventSlowTimeout:        &eventSlowTimeout,
		EventHandlerConcurrency: abxbus.EventHandlerConcurrencyParallel,
		EventHandlerCompletion:  abxbus.EventHandlerCompletionAll,
		EventHandlerSlowTimeout: &handlerSlowTimeout,
	})
	h := bus.On("Evt", "h", func(e *abxbus.BaseEvent, ctx context.Context) (any, error) { return "ok", nil }, nil)
	e := bus.Emit(abxbus.NewBaseEvent("Evt", map[string]any{"k": "v"}))
	if _, err := e.Now(); err != nil {
		t.Fatal(err)
	}

	data, err := bus.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	var payload abxbus.EventBusJSON
	if err := json.Unmarshal(data, &payload); err != nil {
		t.Fatal(err)
	}
	if payload.ID != bus.ID || payload.Name != bus.Name {
		t.Fatalf("id/name mismatch in json payload: %#v", payload)
	}
	if payload.MaxHistorySize == nil || *payload.MaxHistorySize != maxHistory {
		t.Fatalf("max_history_size mismatch in json payload: %#v", payload.MaxHistorySize)
	}
	if !payload.MaxHistoryDrop {
		t.Fatalf("expected max_history_drop=true in json payload")
	}
	if payload.EventConcurrency != abxbus.EventConcurrencyParallel || payload.EventHandlerConcurrency != abxbus.EventHandlerConcurrencyParallel {
		t.Fatalf("concurrency fields mismatch in payload")
	}
	if len(payload.Handlers) != 1 || payload.Handlers[h.ID] == nil {
		t.Fatalf("handler map mismatch in payload: %#v", payload.Handlers)
	}
	if len(payload.HandlersByKey["Evt"]) != 1 || payload.HandlersByKey["Evt"][0] != h.ID {
		t.Fatalf("handlers_by_key mismatch in payload: %#v", payload.HandlersByKey)
	}
	if payload.EventHistory[e.EventID] == nil {
		t.Fatalf("event history missing emitted event id=%s", e.EventID)
	}

	restored, err := abxbus.EventBusFromJSON(data)
	if err != nil {
		t.Fatal(err)
	}
	if restored.ID != bus.ID || restored.Name != bus.Name {
		t.Fatalf("id/name mismatch after roundtrip")
	}
	if restored.EventTimeout == nil || *restored.EventTimeout != eventTimeout {
		t.Fatalf("event timeout mismatch after roundtrip")
	}
	if restored.EventSlowTimeout == nil || *restored.EventSlowTimeout != eventSlowTimeout {
		t.Fatalf("event slow timeout mismatch after roundtrip")
	}
	if restored.EventHandlerSlowTimeout == nil || *restored.EventHandlerSlowTimeout != handlerSlowTimeout {
		t.Fatalf("handler slow timeout mismatch after roundtrip")
	}
	if restored.EventHistory.Size() != 1 {
		t.Fatalf("expected one history entry after roundtrip, got %d", restored.EventHistory.Size())
	}
	restoredEvent := restored.EventHistory.GetEvent(e.EventID)
	if restoredEvent == nil || restoredEvent.Payload["k"] != "v" {
		t.Fatalf("restored history payload mismatch")
	}
	if len(restoredEvent.EventResults) != 1 {
		t.Fatalf("expected one restored event result, got %d", len(restoredEvent.EventResults))
	}
	for _, result := range restoredEvent.EventResults {
		if result.Handler == nil {
			t.Fatalf("restored event result should reference restored handler object")
		}
		if result.HandlerID != result.Handler.ID {
			t.Fatalf("restored handler linkage mismatch")
		}
	}
	if !restored.IsIdleAndQueueEmpty() {
		t.Fatalf("restored idle bus should start with clean runtime state")
	}

	restored.On("Evt2", "h2", func(e *abxbus.BaseEvent, ctx context.Context) (any, error) { return "ok2", nil }, nil)
	v, err := restored.Emit(abxbus.NewBaseEvent("Evt2", nil)).EventResult()
	if err != nil || v != "ok2" {
		t.Fatalf("restored bus should remain functional, result=%#v err=%v", v, err)
	}
}

func TestEventBusSerializationPreservesUnboundedHistoryNull(t *testing.T) {
	unlimitedHistorySize := abxbus.UnlimitedHistorySize
	bus := abxbus.NewEventBus("UnlimitedSerBus", &abxbus.EventBusOptions{
		MaxHistorySize: &unlimitedHistorySize,
		MaxHistoryDrop: false,
	})
	data, err := bus.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Contains(data, []byte(`"max_history_size":null`)) {
		t.Fatalf("expected max_history_size to serialize as null: %s", string(data))
	}

	var payload abxbus.EventBusJSON
	if err := json.Unmarshal(data, &payload); err != nil {
		t.Fatal(err)
	}
	if payload.MaxHistorySize != nil {
		t.Fatalf("expected unmarshaled max_history_size null, got %#v", payload.MaxHistorySize)
	}

	restored, err := abxbus.EventBusFromJSON(data)
	if err != nil {
		t.Fatal(err)
	}
	if restored.EventHistory.MaxHistorySize != nil {
		t.Fatalf("expected restored history to remain unbounded, got %#v", restored.EventHistory.MaxHistorySize)
	}
}

func TestEventBusFromJSONNullEventTimeoutUsesDefault(t *testing.T) {
	data := []byte(`{"id":"timeout-null-bus","name":"TimeoutNullBus","max_history_size":100,"max_history_drop":false,"event_concurrency":"bus-serial","event_timeout":null,"event_slow_timeout":null,"event_handler_concurrency":"serial","event_handler_completion":"all","event_handler_slow_timeout":null,"event_handler_detect_file_paths":false,"handlers":{},"handlers_by_key":{},"event_history":{},"pending_event_queue":[]}`)
	restored, err := abxbus.EventBusFromJSON(data)
	if err != nil {
		t.Fatal(err)
	}
	if restored.EventTimeout == nil || *restored.EventTimeout != 60.0 {
		t.Fatalf("JSON event_timeout:null should use the default bus timeout, got %#v", restored.EventTimeout)
	}
	roundtripped, err := restored.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Contains(roundtripped, []byte(`"event_timeout":60`)) {
		t.Fatalf("expected event_timeout default to survive roundtrip: %s", string(roundtripped))
	}
}

func TestEventBusFromJSONDefaultsMissingHandlerMaps(t *testing.T) {
	data := []byte(`{"id":"missing-handler-maps","name":"MissingHandlerMaps","max_history_size":100,"max_history_drop":false,"event_concurrency":"bus-serial","event_timeout":null,"event_slow_timeout":null,"event_handler_concurrency":"serial","event_handler_completion":"all","event_handler_slow_timeout":null,"event_handler_detect_file_paths":false,"event_history":{},"pending_event_queue":[]}`)
	restored, err := abxbus.EventBusFromJSON(data)
	if err != nil {
		t.Fatal(err)
	}
	restored.On("Evt", "handler", func(event *abxbus.BaseEvent, ctx context.Context) (any, error) {
		return "ok", nil
	}, nil)
	result, err := restored.Emit(abxbus.NewBaseEvent("Evt", nil)).EventResult()
	if err != nil {
		t.Fatal(err)
	}
	if result != "ok" {
		t.Fatalf("restored bus handler did not run after missing maps were defaulted: %#v", result)
	}
}

func TestEventBusSerializationPreservesHandlerRegistrationOrderThroughJSONAndRestore(t *testing.T) {
	detectPaths := false
	bus := abxbus.NewEventBus("HandlerOrderSourceBus", &abxbus.EventBusOptions{
		EventHandlerConcurrency:     abxbus.EventHandlerConcurrencySerial,
		EventHandlerCompletion:      abxbus.EventHandlerCompletionAll,
		EventHandlerDetectFilePaths: &detectPaths,
	})
	originalOrder := []string{}

	first := bus.On("HandlerOrderEvent", "first", func(e *abxbus.BaseEvent, ctx context.Context) (any, error) {
		originalOrder = append(originalOrder, "first")
		return "first", nil
	}, nil)
	second := bus.On("HandlerOrderEvent", "second", func(e *abxbus.BaseEvent, ctx context.Context) (any, error) {
		originalOrder = append(originalOrder, "second")
		return "second", nil
	}, nil)
	expectedIDs := []string{first.ID, second.ID}

	data, err := bus.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	assertJSONKeyBefore(t, data, first.ID, second.ID)
	var payload abxbus.EventBusJSON
	if err := json.Unmarshal(data, &payload); err != nil {
		t.Fatal(err)
	}
	if got := payload.HandlersByKey["HandlerOrderEvent"]; len(got) != 2 || got[0] != expectedIDs[0] || got[1] != expectedIDs[1] {
		t.Fatalf("handlers_by_key order mismatch: got %v want %v", got, expectedIDs)
	}

	if _, err := bus.Emit(abxbus.NewBaseEvent("HandlerOrderEvent", nil)).Now(); err != nil {
		t.Fatal(err)
	}
	if len(originalOrder) != 2 || originalOrder[0] != "first" || originalOrder[1] != "second" {
		t.Fatalf("handler execution order mismatch before restore: got %v", originalOrder)
	}

	restored, err := abxbus.EventBusFromJSON(data)
	if err != nil {
		t.Fatal(err)
	}
	restoredData, err := restored.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	assertJSONKeyBefore(t, restoredData, first.ID, second.ID)
	var restoredPayload abxbus.EventBusJSON
	if err := json.Unmarshal(restoredData, &restoredPayload); err != nil {
		t.Fatal(err)
	}
	if got := restoredPayload.HandlersByKey["HandlerOrderEvent"]; len(got) != 2 || got[0] != expectedIDs[0] || got[1] != expectedIDs[1] {
		t.Fatalf("restored handlers_by_key order mismatch: got %v want %v", got, expectedIDs)
	}

	restoredOrder := []string{}
	restored.On("HandlerOrderEvent", "first", func(e *abxbus.BaseEvent, ctx context.Context) (any, error) {
		restoredOrder = append(restoredOrder, "first")
		return "first", nil
	}, payload.Handlers[first.ID])
	restored.On("HandlerOrderEvent", "second", func(e *abxbus.BaseEvent, ctx context.Context) (any, error) {
		restoredOrder = append(restoredOrder, "second")
		return "second", nil
	}, payload.Handlers[second.ID])

	restoredData, err = restored.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	assertJSONKeyBefore(t, restoredData, first.ID, second.ID)
	if err := json.Unmarshal(restoredData, &restoredPayload); err != nil {
		t.Fatal(err)
	}
	if got := restoredPayload.HandlersByKey["HandlerOrderEvent"]; len(got) != 2 || got[0] != expectedIDs[0] || got[1] != expectedIDs[1] {
		t.Fatalf("reattached handlers_by_key order mismatch: got %v want %v", got, expectedIDs)
	}

	if _, err := restored.Emit(abxbus.NewBaseEvent("HandlerOrderEvent", nil)).Now(); err != nil {
		t.Fatal(err)
	}
	if len(restoredOrder) != 2 || restoredOrder[0] != "first" || restoredOrder[1] != "second" {
		t.Fatalf("handler execution order mismatch after restore: got %v", restoredOrder)
	}
}

func TestEventBusFromJSONRecreatesMissingHandlerEntriesFromEventResultMetadata(t *testing.T) {
	bus := abxbus.NewEventBus("MissingHandlerHydrationBus", nil)
	bus.On("SerializableEvent", "handler", func(e *abxbus.BaseEvent, ctx context.Context) (any, error) {
		return "ok", nil
	}, nil)
	event := bus.Emit(abxbus.NewBaseEvent("SerializableEvent", nil))
	if _, err := event.Now(); err != nil {
		t.Fatal(err)
	}

	handlerID := ""
	for _, result := range event.EventResults {
		handlerID = result.HandlerID
		break
	}
	data, err := bus.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	var payload map[string]any
	if err := json.Unmarshal(data, &payload); err != nil {
		t.Fatal(err)
	}
	payload["handlers"] = map[string]any{}
	payload["handlers_by_key"] = map[string]any{}
	data, err = json.Marshal(payload)
	if err != nil {
		t.Fatal(err)
	}

	restored, err := abxbus.EventBusFromJSON(data)
	if err != nil {
		t.Fatal(err)
	}
	restoredEvent := restored.EventHistory.GetEvent(event.EventID)
	if restoredEvent == nil {
		t.Fatalf("restored event missing")
	}
	restoredData, err := restored.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	var restoredPayload abxbus.EventBusJSON
	if err := json.Unmarshal(restoredData, &restoredPayload); err != nil {
		t.Fatal(err)
	}
	if _, ok := restoredPayload.Handlers[handlerID]; !ok {
		t.Fatalf("restored handlers missing hydrated handler %s", handlerID)
	}
	restoredResult := restoredEvent.EventResults[handlerID]
	if restoredResult == nil || restoredResult.Handler == nil || restoredResult.Handler.ID != handlerID {
		t.Fatalf("restored result handler linkage mismatch: %#v", restoredResult)
	}
}

func TestBaseEventFromJSONRoundtripsRuntimeJSONShape(t *testing.T) {
	bus := abxbus.NewEventBus("SerializableBaseEventBus", nil)
	defer bus.Destroy()
	bus.On("SerializableBaseEvent", "handler", func(e *abxbus.BaseEvent, ctx context.Context) (any, error) {
		return "ok", nil
	}, nil)

	event := bus.Emit(abxbus.NewBaseEvent("SerializableBaseEvent", nil))
	if _, err := event.Now(); err != nil {
		t.Fatal(err)
	}
	payload, err := event.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	restored, err := abxbus.BaseEventFromJSON(payload)
	if err != nil {
		t.Fatal(err)
	}
	restoredPayload, err := restored.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(restoredPayload, payload) {
		t.Fatalf("BaseEvent JSON roundtrip mismatch:\nrestored=%s\noriginal=%s", restoredPayload, payload)
	}
}

func TestEventBusSerializationPreservesPendingQueueIDs(t *testing.T) {
	bus := abxbus.NewEventBus("PendingSerBus", &abxbus.EventBusOptions{EventHandlerConcurrency: abxbus.EventHandlerConcurrencySerial})
	started := make(chan struct{}, 1)
	release := make(chan struct{})
	bus.On("BlockedEvt", "block", func(e *abxbus.BaseEvent, ctx context.Context) (any, error) {
		started <- struct{}{}
		<-release
		return "done", nil
	}, nil)

	first := bus.Emit(abxbus.NewBaseEvent("BlockedEvt", nil))
	<-started
	second := bus.Emit(abxbus.NewBaseEvent("BlockedEvt", nil))

	data, err := bus.ToJSON()
	if err != nil {
		close(release)
		t.Fatal(err)
	}
	var payload abxbus.EventBusJSON
	if err := json.Unmarshal(data, &payload); err != nil {
		close(release)
		t.Fatal(err)
	}
	if len(payload.PendingEventQueue) == 0 {
		close(release)
		t.Fatalf("expected at least one pending event id in serialization payload")
	}
	foundSecond := false
	for _, eventID := range payload.PendingEventQueue {
		if eventID == second.EventID {
			foundSecond = true
			break
		}
	}
	if !foundSecond {
		close(release)
		t.Fatalf("expected queued second event id in pending_event_queue, got %v", payload.PendingEventQueue)
	}

	restored, err := abxbus.EventBusFromJSON(data)
	if err != nil {
		close(release)
		t.Fatal(err)
	}
	restoredData, err := restored.ToJSON()
	if err != nil {
		close(release)
		t.Fatal(err)
	}
	var restoredPayload abxbus.EventBusJSON
	if err := json.Unmarshal(restoredData, &restoredPayload); err != nil {
		close(release)
		t.Fatal(err)
	}
	foundSecondAfterRestore := false
	for _, eventID := range restoredPayload.PendingEventQueue {
		if eventID == second.EventID {
			foundSecondAfterRestore = true
			break
		}
	}
	if !foundSecondAfterRestore {
		close(release)
		t.Fatalf("restored pending_event_queue missing second event id")
	}

	restoredQueued := restored.EventHistory.GetEvent(second.EventID)
	if restoredQueued == nil {
		close(release)
		t.Fatalf("restored history missing queued event")
	}
	if _, err := restoredQueued.Now(); err != nil {
		close(release)
		t.Fatalf("restored queued event should still be processable: %v", err)
	}

	close(release)
	if _, err := first.Now(); err != nil {
		t.Fatal(err)
	}
}

func TestEventBusFromJSONPreservesEventHistoryObjectOrder(t *testing.T) {
	bus := abxbus.NewEventBus("HistoryOrderBus", nil)
	first := abxbus.NewBaseEvent("HistoryOrderEvent", map[string]any{"label": "first"})
	second := abxbus.NewBaseEvent("HistoryOrderEvent", map[string]any{"label": "second"})
	bus.EventHistory.AddEvent(first)
	bus.EventHistory.AddEvent(second)

	data, err := bus.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	assertJSONKeyBefore(t, data, first.EventID, second.EventID)

	restored, err := abxbus.EventBusFromJSON(data)
	if err != nil {
		t.Fatal(err)
	}
	restoredData, err := restored.ToJSON()
	if err != nil {
		t.Fatal(err)
	}
	assertJSONKeyBefore(t, restoredData, first.EventID, second.EventID)
}

File: abxbus-ts/tests/EventBus_serialization.test.ts
import assert from 'node:assert/strict'
import { test } from 'node:test'
import { z } from 'zod'

import { BaseEvent, EventBus } from '../src/index.js'

const delay = (ms: number): Promise<void> =>
  new Promise((resolve) => {
    setTimeout(resolve, ms)
  })

test('EventBus toJSON/fromJSON roundtrip uses id-keyed structures', async () => {
  const bus = new EventBus('SerializableBus', {
    id: '018f8e40-1234-7000-8000-000000001234',
    max_history_size: 500,
    max_history_drop: false,
    event_concurrency: 'parallel',
    event_handler_concurrency: 'parallel',
    event_handler_completion: 'first',
    event_timeout: 0,
    event_handler_slow_timeout: 12,
    event_slow_timeout: 34,
    event_handler_detect_file_paths: false,
  })
  const SerializableEvent = BaseEvent.extend('SerializableEvent', {})

  bus.on(SerializableEvent, async () => {
    await delay(20)
    return 'ok'
  })

  const release_pause = bus.locks._requestRunloopPause()
  const pending_event = bus.emit(SerializableEvent({ event_timeout: 11, event_concurrency: 'bus-serial' } as any))
  await Promise.resolve()

  const json = bus.toJSON()
  assert.equal(json.id, '018f8e40-1234-7000-8000-000000001234')
  assert.equal(json.name, 'SerializableBus')
  assert.equal(Object.keys(json.handlers).length, 1)
  assert.equal(Object.keys(json.handlers_by_key).length, 1)
  assert.equal(Array.isArray(json.handlers_by_key.SerializableEvent), true)
  assert.equal(Object.keys(json.event_history).length, 1)
  assert.equal((json.event_history[pending_event.event_id] as Record<string, unknown>).event_id, pending_event.event_id)
  assert.equal(json.pending_event_queue.length, 1)
  assert.equal(json.pending_event_queue[0], pending_event.event_id)

  const restored = EventBus.fromJSON(json)
  assert.equal(restored.id, '018f8e40-1234-7000-8000-000000001234')
  assert.equal(restored.name, 'SerializableBus')
  assert.equal(restored.event_history.max_history_size, 500)
  assert.equal(restored.event_history.max_history_drop, false)
  assert.equal(restored.event_concurrency, 'parallel')
  assert.equal(restored.event_handler_concurrency, 'parallel')
  assert.equal(restored.event_handler_completion, 'first')
  assert.equal(restored.event_timeout, 0)
  assert.equal(restored.event_handler_slow_timeout, 12)
  assert.equal(restored.event_slow_timeout, 34)
  assert.equal(restored.event_handler_detect_file_paths, false)
  assert.equal(restored.handlers.size, 1)
  assert.equal(restored.handlers_by_key.get('SerializableEvent')?.length, 1)
  assert.equal(restored.event_history.size, 1)
  assert.equal(restored.pending_event_queue.length, 1)
  assert.equal(restored.pending_event_queue[0].event_id, pending_event.event_id)
  assert.equal(restored.runloop_running, false)

  release_pause()
  await pending_event.now()
})

test('EventBus serialization preserves unbounded history null', () => {
  const bus = new EventBus('UnlimitedSerBus', {
    max_history_size: null,
    max_history_drop: false,
  })
  try {
    const json = bus.toJSON()
    assert.equal(json.max_history_size, null)

    const restored = EventBus.fromJSON(json)
    assert.equal(restored.event_history.max_history_size, null)
  } finally {
    bus.destroy()
  }
})

test('EventBus.fromJSON null event_timeout uses default', () => {
  const bus = new EventBus('TimeoutNullBus')
  try {
    const json = bus.toJSON()
    json.event_timeout = null

    const restored = EventBus.fromJSON(json)
    assert.equal(restored.event_timeout, 60)
    assert.equal(restored.toJSON().event_timeout, 60)
  } finally {
    bus.destroy()
  }
})

test('EventBus.fromJSON defaults missing handler maps', async () => {
  const json = new EventBus('MissingHandlerMaps').toJSON()
  delete (json as Partial<typeof json>).handlers
  delete (json as Partial<typeof json>).handlers_by_key

  const restored = EventBus.fromJSON(json)
  const SerializableEvent = BaseEvent.extend('SerializableEvent', {})
  try {
    restored.on(SerializableEvent, () => 'ok')
    const result = await restored.emit(SerializableEvent({})).eventResult()
    assert.equal(result, 'ok')
  } finally {
    await restored.destroy()
  }
})

test('EventBus preserves handler registration order through JSON and restore', async () => {
  const HandlerOrderEvent = BaseEvent.extend('HandlerOrderEvent', {})
  const bus = new EventBus('HandlerOrderSourceBus', {
    event_handler_concurrency: 'serial',
    event_handler_completion: 'all',
    event_handler_detect_file_paths: false,
  })
  const original_order: string[] = []

  const first = bus.on(HandlerOrderEvent, () => {
    original_order.push('first')
    return 'first'
  })
  const second = bus.on(HandlerOrderEvent, () => {
    original_order.push('second')
    return 'second'
  })
  const expected_ids = [first.id, second.id]

  const json = bus.toJSON()
  assert.deepEqual(Object.keys(json.handlers), expected_ids)
  assert.deepEqual(json.handlers_by_key.HandlerOrderEvent, expected_ids)

  await bus.emit(HandlerOrderEvent({})).now()
  assert.deepEqual(original_order, ['first', 'second'])

  const restored = EventBus.fromJSON(json)
  const restored_json = restored.toJSON()
  assert.deepEqual(Array.from(restored.handlers.keys()), expected_ids)
  assert.deepEqual(restored.handlers_by_key.get('HandlerOrderEvent'), expected_ids)
  assert.deepEqual(Object.keys(restored_json.handlers), expected_ids)
  assert.deepEqual(restored_json.handlers_by_key.HandlerOrderEvent, expected_ids)

  const restored_order: string[] = []
  restored.handlers.get(first.id)!.handler = () => {
    restored_order.push('first')
    return 'first'
  }
  restored.handlers.get(second.id)!.handler = () => {
    restored_order.push('second')
    return 'second'
  }

  await restored.emit(HandlerOrderEvent({})).now()
  assert.deepEqual(restored_order, ['first', 'second'])

  bus.destroy()
  restored.destroy()
})

test('EventBus.fromJSON recreates missing handler entries from event_result metadata', async () => {
  const bus = new EventBus('MissingHandlerHydrationBus', {
    event_handler_detect_file_paths: false,
  })
  const SerializableEvent = BaseEvent.extend('MissingHandlerHydrationEvent', {})

  bus.on(SerializableEvent, () => 'ok')
  const event = bus.emit(SerializableEvent({}))
  await event.now()

  const handler_id = Array.from(event.event_results.values())[0].handler_id
  const json = bus.toJSON()
  json.handlers = {}
  json.handlers_by_key = {}

  const restored = EventBus.fromJSON(json)
  const restored_event = restored.event_history.get(event.event_id)
  assert.ok(restored_event)
  assert.ok(restored.handlers.has(handler_id))
  const restored_result = restored_event!.event_results.get(handler_id)
  assert.ok(restored_result)
  assert.equal(restored_result!.handler, restored.handlers.get(handler_id))
  assert.equal(typeof restored_result!.handler.handler, 'function')
  assert.equal(await restored_result!.handler.handler(restored_event as BaseEvent), undefined)
})

test('BaseEvent.fromJSON roundtrips runtime JSON shape', async () => {
  const bus = new EventBus('SerializableBaseEventBus', {
    event_handler_detect_file_paths: false,
  })
  const SerializableEvent = BaseEvent.extend('SerializableBaseEvent', {})
  try {
    bus.on(SerializableEvent, () => 'ok')
    const event = bus.emit(SerializableEvent({}))
    await event.now()

    const payload = event.toJSON()
    const restored_payload = BaseEvent.fromJSON(payload).toJSON()
    assert.deepEqual(restored_payload, payload)
  } finally {
    await bus.destroy()
  }
})

test('EventBus toJSON promotes pending events into event_history snapshot', async () => {
  const bus = new EventBus('ModelDumpPendingBus')
  const PendingEvent = BaseEvent.extend('ModelDumpPendingEvent', {})

  bus.on(PendingEvent, async () => {
    await delay(10)
    return 'ok'
  })

  const release_pause = bus.locks._requestRunloopPause()
  const pending = bus.emit(PendingEvent({}))
  await Promise.resolve()

  const json = bus.toJSON()
  assert.equal(Boolean(json.event_history[pending.event_id]), true)
  assert.equal(json.pending_event_queue.includes(pending.event_id), true)

  release_pause()
  await pending.now()
})

test('EventBus toJSON serializes history events when Zod event schema encode fails', () => {
  const bus = new EventBus('TransformHistorySerializationBus')
  const TransformEvent = BaseEvent.extend('TransformHistorySerializationEvent', {
    count: z.string().transform(Number),
  })
  try {
    const event = TransformEvent({ count: '7' })
    bus.event_history.set(event.event_id, event)

    assert.equal((Array.from(bus.event_history.values())[0] as unknown as { count: number }).count, 7)
    const json = bus.toJSON()
    assert.equal((json.event_history[event.event_id] as Record<string, unknown>).count, 7)
  } finally {
    bus.destroy()
  }
})

test('EventBus toJSON falls back to closest JSON Schema for transform event_result_type fields', () => {
  const bus = new EventBus('TransformResultSchemaSerializationBus')
  const TransformResultEvent = BaseEvent.extend('TransformResultSchemaSerializationEvent', {
    event_result_type: z.object({
      label: z.string(),
      count: z.string().transform(Number),
      nested: z.object({
        ok: z.boolean(),
        bad: z.string().transform(Number),
      }),
      maybe: z.string().transform(Number).optional(),
      items: z.array(z.string().transform(Number)),
      choice: z.union([z.string().transform(Number), z.number()]),
    }),
  })
  try {
    const event = TransformResultEvent({})
    bus.event_history.set(event.event_id, event)

    const json = bus.toJSON()
    const result_schema = (json.event_history[event.event_id] as Record<string, unknown>).event_result_type as Record<string, unknown>
    const properties = result_schema.properties as Record<string, unknown>
    assert.deepEqual(Object.keys(properties).sort(), ['choice', 'count', 'items', 'label', 'maybe', 'nested'])
    assert.deepEqual(result_schema.required, ['choice', 'count', 'items', 'label', 'nested'])
    assert.equal((properties.label as Record<string, unknown>).type, 'string')
    assert.equal((properties.count as Record<string, unknown>).type, 'string')
    assert.equal((properties.maybe as Record<string, unknown>).type, 'string')
    assert.equal(((properties.items as Record<string, unknown>).items as Record<string, unknown>).type, 'string')
    assert.deepEqual((properties.choice as Record<string, unknown>).anyOf, [{ type: 'string' }, { type: 'number' }])

    const nested_properties = (properties.nested as Record<string, unknown>).properties as Record<string, unknown>
    assert.deepEqual(Object.keys(nested_properties).sort(), ['bad', 'ok'])
    assert.deepEqual((properties.nested as Record<string, unknown>).required, ['bad', 'ok'])
    assert.equal((nested_properties.bad as Record<string, unknown>).type, 'string')
  } finally {
    bus.destroy()
  }
})

test('EventBus.fromJSON preserves event_history object order', () => {
  const bus = new EventBus('HistoryOrderBus')
  const HistoryOrderEvent = BaseEvent.extend('HistoryOrderEvent', {})
  try {
    const first = HistoryOrderEvent({})
    const second = HistoryOrderEvent({})
    bus.event_history.set(first.event_id, first)
    bus.event_history.set(second.event_id, second)

    const json = bus.toJSON()
    assert.deepEqual(Object.keys(json.event_history), [first.event_id, second.event_id])

    const restored = EventBus.fromJSON(json)
    assert.deepEqual(Object.keys(restored.toJSON().event_history), [first.event_id, second.event_id])
  } finally {
    bus.destroy()
  }
})
